### DISCUSSION TITLE
TUI freeze

### DISCUSSION THREAD

**Discussion by pom11** [Textualize/textual#368] [General]
Hello!
Given the following widget for displaying docker containers stats i found out that the TUI is freezing for about 2 seconds at a `refresh_rate` of 10 seconds. Is this normal behaviour or will be solved in future updates? 

Note: Docker environment has 32CPU 64GB RAM and there where 10 containers running.

```python
from python_on_whales import docker as pwhales

class DockerContainerStats(Widget):

	def __init__(self, refresh_rate: int) -> None:
		self.refresh_rate = refresh_rate
		super().__init__()

	mouse_over = Reactive(False)

	def on_mount(self):
		self.set_interval(self.refresh_rate, self.refresh)

	def getContainers(self) -> Table:

		containers = pwhales.stats()

		table = Table(show_header=False, show_lines=False, box=box.HEAVY)
		table.add_column("Container", no_wrap=False, justify="left") 
		table.add_column("Name", no_wrap=False, justify="left") 
		table.add_column("CPU %", no_wrap=False, justify="left") 
		table.add_column("MEM %", no_wrap=False, justify="left") 

		for c in containers:
			table.add_row('[bold cyan]'+str(c.container), '[bold cyan]'+str(c.container_name), '[bold cyan]'+str(c.cpu_percentage), '[bold cyan]'+str(c.memory_percentage) )
		return table

	def render(self):
		table = self.getContainers()
		return Static(renderable=table)
```

**Comment by willmcgugan** [comment]
Have you done any profiling? How long does `stats` take to run? And how big is the table?

In general, it's not a good idea to do more than return a renderable in your `render` method. i.e. have another interval or test to retrieve the data, which render returns.

**Reply by pom11** [reply]
3 seconds to display the TUI. The table has 4 columns with 10 rows (+ headers). The only profiling i can do during runtime is to watch the included time of TUI (header). I dont understand your recommendation, sorry. Should i add all the `docker stats` logic in `App` and pass them to the `Widget` as i did with `refresh_rate` interval?

**Reply by pom11** [reply]
3 seconds at startup of TUI. 2 seconds per refresh. In interpreter the `containers = pwhales.stats()` takes 2 seconds too. I thought the TUI wont freeze until refresh is done.
Didn't looked at the source code of `textual` but i think there is a `await` problem/misconfig.

**Reply by willmcgugan** [reply]
It's possible there is a performance issue in Rich or Textual, but it is unlikely. What is more likely is that your call to `pwhales.stats()` is going to the network and taking a while to return, during which time Textual is unable to update the screen.

You will likely need to run `pwhales.stats` in a thread. The `concurrent.futures` module can help you with this.

You can confirm this by timing how long the `pwhales.stats` call takes to execute. You could use `self.log` to write the times to the log file.

**Reply by pom11** [reply]
Understood. That would be a good update for Textual to use a thread for a widget by default. Thx for info. Will get back with updates

**Reply by kfmfe04** [reply]
If you do get this working with `futures`, please share your code.  I'm curious as to what this pattern looks like in the context of `Textual`.  Thank you!
